src/lib.rs

```diff
@@ -19,27 +19,60 @@
+//! while handling such loop-stage events.
+//!
+//! Every event loop starts with a [`NewEvents`] event with [`StartCause::Init`], followed by a [`Resumed`] loop-stage
+//! event. Application initialization - including initializing graphics resources - should typically be done while handling
+//! this first [`Resumed`] event.
+//!
+//! The primary loop of events always starts with the loop-stage event [`NewEvents`], containing a [`StartCause`].
+//! This is then followed by any "main events" generated by the platform. "Main events" include any and all
+//! [`WindowEvent`]s, [`DeviceEvent`]s and [`UserEvent`]s, but not any redraw events. The [`MainEventsCleared`] loop-stage event
+//! indicates when all main events have been exhausted and handling this is typically where the "main body" of your application
+//! should be run, since all input has been handled for this loop iteration. Applications which need to constantly update
+//! their graphics (for example, games) will typically call [`request_redraw()`] while handling [`MainEventsCleared`].
+//!
+//! If the window needs to be redrawn (either because the platform requests it, or application code has), then a
+//! [`RedrawRequested`] event follows. winit will ensure that this is sent at most once so that applications can run rendering code here,
+//! confident that work will not be duplicated. This is always followed by the [`RedrawEventsCleared`] loop-stage event,
+//! marking the end of the primary loop.
+//!
+//! Some platforms currently run one iteration of this primary loop immediately after the first [`Resumed`] event, but
+//! this is platform-dependent and may change.
+//!
+//! The [`Suspended`] and [`Resumed`] loop-stage events are used on some platforms (typically mobile platforms) to
+//! indicate suspension and resumption of the application by the platform. See the [detailed docs][`Suspended`] for details.
+//!
+//! Once the [`ControlFlow`] has been set to exit (typically with [`ControlFlow::set_exit()`] or [`set_exit_with_code()`][`ControlFlow::set_exit()`])
+//! a final [`LoopDestroyed`] event is dispatched. This is irreversible and a typical place to free all application resources,
+//! as the process will be terminated after this.
+//!
+//! This overview should be correct for cross-platform use in most cases, please see the detailed descriptions in the
+//! API docs for the exact behaviour of each event and details of any platform-dependent behaviour. In particular on some
+//! platforms it is possible (though discouraged) to use [`run_return()`][`EventLoopExtRunReturn::run_return`] to prevent
+//! [`run()`][`EventLoop::run()`] from hijacking the process, however this may change the order of events significantly,
+//! so carefully check the API documentation.
+//!
+//! The [`winit_input_helper`](https://crates.io/crates/winit_input_helper) crate can help the mangement of main events
```

Comment:
While I generally like suggesting crates to do a specific task, it seems like quite a prominent place to put it? If our event handling is so difficult to understand that we need to direct users to a different crate, perhaps we should adopt that crate's API instead?

Relatedly, I suspect the rendering logic presented in the readme in `winit_input_helper` is wrong.

Reply:
Absolutely right, this wasn't appropriate.

Reply:
No worries, you were just trying to make the library better for everyone!